### PR TITLE
RavenDB-18413 Timings aren't filled after ToQueryable

### PR DESCRIPTION
**Pull request by shaharhikri**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-18413

### Additional description

Fixing timings aren't filled after ToQueryable

### Type of change

- Bug fix

### How risky is the change?

- Low

### Backward compatibility

- Not relevant

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing 

- Tests have been added that prove the fix is effective or that the feature works

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

### CODE REVIEW

**Comment by shaharhikri** [comment]
Checkin failing tests.

File: src/Raven.Client/Documents/Session/DocumentQuery.cs [hunk 1]
```diff
@@ -1213,6 +1213,7 @@ public IRavenQueryable<T> ToQueryable()
 
             var queryStatistics = new QueryStatistics();
             var highlightings = new LinqQueryHighlightings();
+            IncludeTimings(out _);
```
Comment (ayende) [inline src/Raven.Client/Documents/Session/DocumentQuery.cs]: You are doing this unconditionally, which change the meaning, now all queries with `ToQueryable` will have `Timings`, even if they aren't actually in use.

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Raven.Client/Documents/Session/DocumentQuery.cs (hunk 1 above)
Comment (shaharhikri) [reply src/Raven.Client/Documents/Session/DocumentQuery.cs]: Fixed.

File: src/Raven.Client/Documents/Session/AbstractDocumentQuery.Timings.cs [hunk 2]
```diff
@@ -6,15 +6,22 @@ public abstract partial class AbstractDocumentQuery<T, TSelf>
     {
         protected QueryTimings QueryTimings;
 
+        protected FlagHolder ShouldIncludeTimings = new FlagHolder { Value = false };
```
Comment (ayende) [inline src/Raven.Client/Documents/Session/AbstractDocumentQuery.Timings.cs]: Why is this needed?

File: src/Raven.Client/Documents/Session/AbstractDocumentQuery.cs [hunk 3]
```diff
@@ -1250,7 +1250,7 @@ private void BuildInclude(StringBuilder queryText)
             if (DocumentIncludes.Count == 0 &&
                 HighlightingTokens.Count == 0 &&
                 ExplanationToken == null &&
-                QueryTimings == null &&
+                !ShouldIncludeTimings.Value &&
```
Comment (ayende) [inline src/Raven.Client/Documents/Session/AbstractDocumentQuery.cs]: Pretty sure that this won't work. Are you using the _same_ instance here across the board?
Because if I'm reading this properly, the instance you use and the internal instance aren't the same.

File: src/Raven.Client/Documents/Session/AbstractDocumentQuery.Timings.cs (hunk 2 above)
Comment (shaharhikri) [reply src/Raven.Client/Documents/Session/AbstractDocumentQuery.Timings.cs]: This property has to be nonprimitive (with reference), because if you update the bool in the original DocumentQuery you want it'll be updated on all of its existed clones (They all need to hold the same instance in 'ShouldIncludeTimings' property).
'bool' is primitive and changing it in the original DocumentQuery won't cause changing it in the clones.

File: src/Raven.Client/Documents/Session/AbstractDocumentQuery.cs (hunk 3 above)
Comment (shaharhikri) [reply src/Raven.Client/Documents/Session/AbstractDocumentQuery.cs]: I use the same instance, that's why I used 'FlagHolder' instead of bool,
when I clone DocumentQuery I copy it.

Comment (ayende) [reply src/Raven.Client/Documents/Session/AbstractDocumentQuery.cs]: I meant, the same instance of `QueryTimings`

Comment (shaharhikri) [reply src/Raven.Client/Documents/Session/AbstractDocumentQuery.cs]: I don't want to use the same instance of 'QueryTimings', because every clone of the original DocumentQuery should have its own QueryTiming, It only needs to behave like the original - if the original includes 'QueryTimings' it should include 'QueryTimings' too.

**Comment by shaharhikri** [comment]
"Assert.QueryTimings" is deprecated, I used "Assert.Same" instead.
image

File: test/SlowTests/Issues/RavenDB-18413.cs [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+namespace SlowTests.Issues
+{
+    public class RavenDB_18413 : RavenTestBase
+    {
+        public RavenDB_18413(ITestOutputHelper output) : base(output)
+        {
+        }
+
+        [Fact]
+        public async Task ToQueryableTimings()
```
Comment (grisha-kotler) [inline test/SlowTests/Issues/RavenDB-18413.cs]: void

File: test/SlowTests/Issues/RavenDB-18413.cs [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+                        .ToQueryable()
+                        .Customize(x => x.Timings(out timings)).ToList();
+
+                    Assert.NotNull(timings.Timings);
+                }
+            }
+        }
+
+        [Fact]
+        public async Task ToQueryableTimingsOutTimings()
```
Comment (grisha-kotler) [inline test/SlowTests/Issues/RavenDB-18413.cs]: void

**Review by grisha-kotler — CHANGES_REQUESTED** [review]

File: test/SlowTests/Issues/RavenDB-18413.cs (hunk 4 above)
Comment (ayende) [reply test/SlowTests/Issues/RavenDB-18413.cs]: Huh? We _never_ want `async void` methods

Comment (grisha-kotler) [reply test/SlowTests/Issues/RavenDB-18413.cs]: Just `void`
```suggestion
        public void ToQueryableTimings()
```

Comment (shaharhikri) [reply test/SlowTests/Issues/RavenDB-18413.cs]: Done

File: test/SlowTests/Issues/RavenDB-18413.cs (hunk 5 above)
Comment (shaharhikri) [reply test/SlowTests/Issues/RavenDB-18413.cs]: Done

File: src/Raven.Client/Documents/Session/AbstractDocumentQuery.Timings.cs (hunk 2 above)
Comment (grisha-kotler) [reply src/Raven.Client/Documents/Session/AbstractDocumentQuery.Timings.cs]: @shaharhikri, let's put the `bool` inside `QueryTimings` and than you don't need the `FlagHolder`

**Review by grisha-kotler — CHANGES_REQUESTED** [review]

**Review by ppekrol — APPROVED** [review]
